### ISSUE TITLE
Dollar vector

### ISSUE THREAD

**Issue by veretin** [vermaseren/form#568]
Hello,
I have problem with the following code:

```
v k,k1,k2, newvector;
c A,B,C;

g f = A(k1)*C(3*k1.k1 + 5*k1.k2 + 7*k2.k2);

* set $k                                                                                                                                
id A(k?$k) = A(k);

  argument C;
  print "in term %t we have $k = %$", $k;
  id $k = newvector;*  <-- this substitution is ignored!
*  id $k.$k = 0;*  <-- this gives error                                                                                                 
*  id $k.k2 = 0;*  <-- this gives error too                                                                                             
  endargument;

print+s;
.end
```

The substitution in "id $k = newvector;" is ignored. 
By the way, the same code with symbols works.

Best regards,
OVeretin

**Comment by jodavies** [comment]
As a quick workaround, does it suffice to use `multiply replace_($k,newvector);` inside the argument environment?

**Comment by veretin** [comment]
No, I want to do more complicated things like

```
id $k.$k^j? = j * $k.$k^(j-1) * $k.V;
```

**Comment by jodavies** [comment]
In that case one could hopefully do something like:
```
multiply replace_($k,tmp);
id tmp.tmp^j? = ... ;
multiply replace_(tmp,$k);
```
but I will take a look at the direct dollar-var version when I get a chance.

**Comment by tueda** [comment]
Another workaround:
```form
id p.p^j? * replace_(p,$k) = j * p.p^(j-1) * p.V * replace_(p,$k);
```
where `p` is a vector.

**Comment by jodavies** [comment]
I spent some time looking at this but did not find exactly where to fix things yet. There are at least two separate issues. The first id statement `id $k = newvector;` is fine on its LHS; it works for symbols as @veretin said, and also for loose vectors (i.e. `k1*k1` is replaced with `newvector*newvector`). In the `k1*k1` case the replacement is made (I think) in `FindRest`. The commentary for that function says "This routine scans for anything but dotproducts and symbols." though, so I suppose in the dotproduct case the pattern should be found elsewhere: @vermaseren do you have any tips here? Where should this pattern match be found?

The second case (`id $k.$k = 0;`) doesn't even make it through the tokenizer ("Illegal second element in dotproduct). If I get it through there (possibly incorrectly) then we get instead "Internal error in code generator. Unknown object: -17" so this pattern is basically "unimplemented" as I understand it.

**Comment by tueda** [comment]
I think `$x.p` (dot product) and `$x(1)` (function) etc. are not implemented for LHS. Such use of dollars needs big changes.

**Comment by jodavies** [comment]
Right, then I think the `id $k = newvector;` case should be considered a bug to be fixed in 4.3.2 and 5.0, and everything else is "wontfix"?

**Comment by tueda** [comment]
Yes, I agree with it. `$x.p` and `$x(1)` could be "wontfix" for now, or we can create a separate issue to address them.
